keps/sig-apps/2214-indexed-job/README.md

```diff
@@ -0,0 +1,566 @@
+
+## Proposal
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a Job author, I can create an array Job where each Pod receives an ordered
+completion index. I can use the index in my binary through an environment
+variable or a file to statically select the load the Pod should work on.
+
+```yaml
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: parallel-work
+spec:
+  completions: 100
+  parallelism: 100
+  template:
+    spec:
+      containers:
+      - name: task
+        image: registry.example.com/processing-image
+        command: ["./process",  "--index", "$INDEX"]
+        env:
+        - name: INDEX
+          valueFrom:
+            fieldRef:
+              fieldPath: metadata.annotations['batch.alpha.kubernetes.io/job-completion-index'] 
+```
+
+### Notes/Constraints/Caveats (Optional)
+
+* An earlier proposal for [indexed Job]
+suggested the support for work lists, i.e. passing different parameters to each
+Pod. We decided to leave this out of the proposal to keep it simple and
+because work lists can be implemented in a startup script using the completion
+index as building block.
+* The semantics of an indexed Job are similar to a StatefulSet, in the sense
+that Pods have an associated index.
+However, the APIs have a major difference: a StatefulSet doesn't have completion
+semantics, as opposed to Jobs.
+
+[indexed Job]: https://github.com/kubernetes/community/blob/b21d1b27c8c748bf81283c2d89cde2becb5f2709/contributors/design-proposals/apps/indexed-job.md
+
+### Risks and Mitigations
+
+Jobs have a known issue in which more than one Pod can be started even if
+parallelism and completion are set to 1 ([reference]). In the case of indexed
+Jobs, this translates to more than one Pod having the same index.
+
+Just like for existing Job patterns, workloads have to handle duplicates at the
+application level.
+
+[reference]: https://kubernetes.io/docs/concepts/workloads/controllers/job/#handling-pod-and-container-failures
+
+## Design Details
+
+### JobSpec API
+
+The JobSpec gets the field `completionMode` to control whether the Job should	
+be treated as an Indexed Job.	
+
+```golang	
+type JobSpec struct {	
+  ...	
+  // CompletionMode specifies how Pod completions are tracked. It can be	
+  // `NonIndexed` (default) or `Indexed`.	
+  // `NonIndexed` means that each Pod completion is homologous to each other.	
+  // The Job is considered complete when there have been .spec.completions	
+  // successful completions.	
+  // `Indexed` means that each Pod completion needs to be tracked individually;	
+  // each Pods gets an associated completion index, which is available in the
+  // annotation	`batch.alpha.kubernetes.io/job-completion-index`.	
+  // The Job is considered complete when there is one successful Pod for each	
+  // index in the range 0 to (.spec.completions - 1).	
+  // When value is `Indexed`, .spec.completions must have a non-zero positive	
+  // value and `.spec.parallelism` must be less than or equal to 10^6.
```

Comment:
I'm assuming that the reasoning behind this 10^6 is the underlying bitmask, right?
i.e. we're saying that tracking 10^6 items in bitmask takes O(125kB) and this is fine, right?

That sounds fine to me, but I would like to ensure that somewhere in the design section we will add a requirement that such big objects won't be updated frequently (e.g. not more often that once per couple seconds or something like that).

Reply:
Yes, this is about being able to maintain the bitmask. Although, if we are going to use it to track completion, we actually need to cap .spec.completions too. The idea is that in the future we don't depend on lingering pods. @erictune thoughts?

> e.g. not more often that once per couple seconds or something like that

Should we be doing some kind of client side throttling? In the current design, Job status updates would come after removing finalizers from finishing Pods, which already throttles Job updates in a way. But if one Pod finishes every 100ms (for example), we might still do more Job updates than what you are suggesting.

Reply:
I changed the limit to be on completions, instead of parallelism